app/src/main/java/ru/practicum/android/diploma/ui/VacancyAdapter.kt

```diff
@@ -1,28 +1,70 @@
 package ru.practicum.android.diploma.ui
 
+import android.util.Log
 import android.view.LayoutInflater
 import android.view.ViewGroup
 import androidx.recyclerview.widget.RecyclerView.Adapter
 import ru.practicum.android.diploma.R
-import ru.practicum.android.diploma.domain.models.Vacancy
+import ru.practicum.android.diploma.ui.search.VacancyListItemUiModel
+import ru.practicum.android.diploma.ui.viewholder.EmptyViewHolder
+import ru.practicum.android.diploma.ui.viewholder.ListItemViewHolder
+import ru.practicum.android.diploma.ui.viewholder.LoadingViewHolder
+import ru.practicum.android.diploma.ui.viewholder.VacancyViewHolder
 import ru.practicum.android.diploma.util.OnItemClickListener
 
-class VacancyAdapter(val listOfVacancies: List<Vacancy>, private val onItemClickListener: OnItemClickListener) :
-    Adapter<VacancyViewHolder>() {
+private const val VIEW_TYPE_EMPTY = 0
+private const val VIEW_TYPE_VACANCY = 1
+private const val VIEW_TYPE_LOADING = 2
 
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VacancyViewHolder {
-        return VacancyViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.vacancy_list_item, parent, false))
-    }
+class VacancyAdapter(private val onItemClickListener: OnItemClickListener) : Adapter<ListItemViewHolder>() {
 
-    override fun getItemCount(): Int {
-        return listOfVacancies.size
+    private val listData = mutableListOf<VacancyListItemUiModel>()
+
+    fun setData(newListData: List<VacancyListItemUiModel>) {
+        listData.clear()
+        listData.addAll(newListData)
+        notifyDataSetChanged()
+
+        Log.d("adapter", "vacanciesListItemUiModel = $itemCount")
     }
 
-    override fun onBindViewHolder(holder: VacancyViewHolder, position: Int) {
-        holder.bind(listOfVacancies[position])
+    override fun getItemViewType(position: Int) =
+        when (listData[position]) {
+            VacancyListItemUiModel.Empty -> VIEW_TYPE_EMPTY
+            VacancyListItemUiModel.Loading -> VIEW_TYPE_LOADING
+            is VacancyListItemUiModel.VacancyItem -> VIEW_TYPE_VACANCY
+        }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = when (viewType) {
+        VIEW_TYPE_VACANCY -> {
+            VacancyViewHolder(
+                LayoutInflater.from(parent.context).inflate(R.layout.vacancy_list_item, parent, false)
+            ) { vacancy -> onItemClickListener.onItemClick(vacancy) }
+        }
+
+        VIEW_TYPE_EMPTY -> {
+            EmptyViewHolder(
+                LayoutInflater.from(parent.context).inflate(R.layout.empty_list_item, parent, false)
+            )
```

Comment:
Тут мне кажется подойдут анонимный классы, ведь логики в них никакой нет
```suggestion
        VIEW_TYPE_EMPTY -> object : ListItemViewHolder(
            LayoutInflater.from(parent.context).inflate(R.layout.empty_list_item, parent, false)
        ) {}
```